client/engine/engine.go

```diff
@@ -245,8 +247,55 @@ func (e *Engine) constructObjectiveFromMessage(message protocols.Message) (proto
 			message.SignedStates[0].State(),
 			*e.store.GetAddress(),
 		)
+	case strings.Contains(string(message.ObjectiveId), "Virtual"):
+		initialState := message.SignedStates[0].State()
+
+		if initialState.TurnNum != 0 {
+			return virtualfund.VirtualFundObjective{}, errors.New("cannot construct virtual fund objective without prefund state")
+		}
+		participants := message.SignedStates[0].State().Participants
+		if len(participants) != 3 {
+			return virtualfund.VirtualFundObjective{}, errors.New("a single hop virtual channel must have exactly 3 participants")
+		}
+
+		alice := participants[0]
+		intermediary := participants[1]
+		bob := participants[2]
+		myAddress := *e.store.GetAddress()
+
+		var left *channel.TwoPartyLedger
+		var right *channel.TwoPartyLedger
+		var ok bool
+
+		if alice != myAddress {
+			left, ok = e.store.GetTwoPartyLedger(intermediary, bob)
+			if !ok {
+				return virtualfund.VirtualFundObjective{}, fmt.Errorf("could not find a left ledger channel between %v and %v", alice, intermediary)
+			}
+		}
+		if bob != myAddress {
+			right, ok = e.store.GetTwoPartyLedger(alice, intermediary)
+			if !ok {
+				return virtualfund.VirtualFundObjective{}, fmt.Errorf("could not find a right ledger channel between %v and %v", intermediary, bob)
+			}
+		}
+
+		myRole, err := getMyRole(myAddress, participants)
+		if err != nil {
+			return virtualfund.VirtualFundObjective{}, errors.New("could not determine my role: %w")
+		}
+
+		return virtualfund.New(
+			true, // TODO ensure objective in only approved if the application has given permission somehow
+			message.SignedStates[0].State(),
```

Comment:
There's a lot of `message.SignedStates[0].State()` in this file, which could be tidied up using the variable you defined on L251. Also, it seems bad that we just index into the 0th state on the message. Could we refactor this away some how? Do we need a `createObjectiveFromState` function instead?

Reply:
This is slightly addressed in [a59c047](https://github.com/statechannels/go-nitro/pull/274/commits/a59c0475342cd93ec949acb4e068812855b853b3) but we should probably extend it to something like `createObjectiveFromState`